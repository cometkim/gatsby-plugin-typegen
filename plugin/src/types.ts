export type { PluginOptions } from './internal/pluginOptions';

/**
 * Copy of the `IGatsbyResolverContext`
 * @see https://github.com/gatsbyjs/gatsby/blob/f10eb58/packages/gatsby/src/schema/type-definitions.ts
 *
 * @todo Provide more accurate type definition.
 */
export interface GatsbyResolverContext {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  nodeModel: any;
}
