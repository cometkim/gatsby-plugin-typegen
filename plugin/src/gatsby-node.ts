import type { GatsbyNode, Reporter } from 'gatsby';
import { interpret } from 'xstate';

import type { TypegenIntepreter } from './services/typegenMachine';
import { makeTypegenMachine } from './services/typegenMachine';
import { makeEmitSchemaService } from './services/emitSchema';
import { makeEmitPluginDocumentService } from './services/emitPluginDocuments';
import type { PluginOptions } from './internal/types';
import { normalizePluginOptions } from './internal/config';
import type { TypegenReporter } from './internal/reporter';

import type { GatsbyStore } from './gatsby-utils';

let service: TypegenIntepreter;

function makeTypegenReporter(reporter: Reporter): TypegenReporter {
  const prefix = '[typegen]';
  return {
    ...reporter,
    log: message => void reporter.log([prefix, message].join(' ')),
    info: message => void reporter.info([prefix, message].join(' ')),
    warn: message => void reporter.warn([prefix, message].join(' ')),
    error: message => void reporter.error([prefix, message].join(' ')),
    verbose: message => void reporter.verbose([prefix, message].join(' ')),
  };
}

export const pluginOptionsSchema: GatsbyNode['pluginOptionsSchema'] = ({
  Joi,
}) => {
  const documentOutputOptionsSchema = Joi.object({
    format: Joi.string()
      .valid('introspection', 'sdl')
      .default('sdl')
      .required(),
    commentDescriptions: Joi.boolean()
      .default(true),
  }).required();

  return Joi.object({
    language: Joi.string()
      .valid('typescript', 'flow')
      .default('typescript'),
    namespace: Joi.string()
      .default('GatsbyTypes'),
    outputPath: Joi.string()
      .default('src/__generated__/gatsby-types.d.ts'),
    includeResolvers: Joi.boolean()
      .default(false),
    autoFix: Joi.boolean()
      .default(true),
    scalars: Joi.object()
      .pattern(/\w[\w\d\-_]+/, Joi.string().required())
      .default({}),
    emitSchema: Joi.object()
      .pattern(/.+/, [
        Joi.boolean().required(),
        documentOutputOptionsSchema,
      ]),
    emitPluginDocuments: Joi.object()
      .pattern(/.+/, [
        Joi.boolean().required(),
        documentOutputOptionsSchema,
      ]),
    emitPluginDocument: Joi.object()
      .pattern(/.+/, [
        Joi.boolean().required(),
        documentOutputOptionsSchema,
      ]),
  });
};

export const onPreBootstrap: GatsbyNode['onPreBootstrap'] = ({
  emitter,
  store: _store,
  reporter: _reporter,
}, options) => {
  const store = _store as GatsbyStore;
  const reporter = makeTypegenReporter(_reporter);

  const { program } = store.getState();
  const basePath = program.directory;

  const pluginOptions = normalizePluginOptions(
    options as PluginOptions,
    { basePath, reporter },
  );

  reporter.verbose(reporter.stripIndent`
    Loaded normalized configuration
    ${JSON.stringify(pluginOptions, null, 2)}
  `);

  const delay = (ms: number) => new Promise(res => setTimeout(res, ms));
  service = interpret(
    makeTypegenMachine({
      context: {
        reporter,
        schema: null,
        trackedDefinitions: null,
        thirdpartyFragmentDefinitions: [],
        hasSchemaChanged: false,
        definitionChangesets: [],
      },
      services: {
        codegen() {
          return delay(1000);
        },
        emitSchema: makeEmitSchemaService({
          pluginOptions,
        }),
        emitPluginDocuments: makeEmitPluginDocumentService({
          pluginOptions,
        }),
        autoFixCodes() {
          return delay(2000);
        },
      },
    }),
  );

  service.start();

  service.onTransition(({ event, value, changed }) => {
    reporter.verbose(`on ${event.type}`);
    if (changed) {
      reporter.verbose(` ⤷ transition to ${JSON.stringify(value)}`);
    } else {
      reporter.verbose(' ⤷ skipped');
    }
  });

  emitter.on('SET_SCHEMA', () => {
    service.send({ type: 'SET_SCHEMA', schema: store.getState().schema });
  });

  emitter.on('SET_GRAPHQL_DEFINITIONS', () => {
    service.send({ type: 'SET_GRAPHQL_DEFINITIONS', definitions: [...store.getState().definitions.values()] });
  });
};

export const onCreateDevServer: GatsbyNode['onCreateDevServer'] = () => {
  service.send('CREATE_DEV_SERVER');
};
