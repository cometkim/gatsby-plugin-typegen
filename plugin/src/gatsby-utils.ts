import path from 'path';
import type { Store } from 'redux';
import type {
  IGatsbyState as GatsbyState,
  ActionsUnion as GatsbyAction,
  IDefinitionMeta as _IDefinitionMeta,
} from 'gatsby/dist/redux/types';
import type { OverrideProps } from '@cometjs/core';

export type GatsbyStore = Store<GatsbyState, GatsbyAction>;

export type IDefinitionMeta = OverrideProps<_IDefinitionMeta, {
  // Trust me, this is more accurate.
  printedAst: string | null,
}>;
export type FragmentDefinition = OverrideProps<IDefinitionMeta, {
  __BRAND__: 'FragmentDefinition',
  isFragment: true,
}>;
export type QueryDefinition = OverrideProps<IDefinitionMeta, {
  __BRAND__: 'QueryDefinition',
  isFragment: false,
}>;

export function isFragmentDefinition(def: IDefinitionMeta): def is FragmentDefinition {
  return def.isFragment;
}

export function isQueryDefinition(def: IDefinitionMeta): def is QueryDefinition {
  return !def.isFragment;
}

/**
 * return `true` if the given definition is assumed to be generated from unnamed query.
 */
export function guessIfUnnnamedQuery(def: QueryDefinition): boolean {
  const { name, filePath } = def;
  const baseDir = path.dirname(filePath);
  return name.startsWith(baseDir.split('/').join(''));
}

export function guessIfThirdPartyDefinition(def: IDefinitionMeta): boolean {
  const { filePath } = def;
  return /(node_modules|\.yarn|\.cache)/.test(filePath);
}

// from https://github.com/gatsbyjs/gatsby/blob/6b4b7f81ec/packages/gatsby/src/schema/print.js#L33-L48
export const gatsbyInternalScalars = [
  'Boolean',
  'Buffer',
  'Date',
  'Float',
  'ID',
  'Int',
  'Internal',
  'InternalInput',
  'JSON',
  'Json',
  'Node',
  'NodeInput',
  'Query',
  'String',
];
