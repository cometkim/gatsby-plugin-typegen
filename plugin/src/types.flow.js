/* @flow */

export type PluginOptions = {

  /**
   * Prefer TypeScript by default, but you can switch to flow.
   *
   * @default 'typescript'
   */
  language?: 'typescript' | 'flow',

  /**
   * Namespace is required to avoid conflict on generated types and other global declarations.
   *
   * Flow will use $ prefix as fallback.
   * e.g) `type GatsbyTypes$MySiteQuery = ...`
   *
   * @default 'GatsbyTypes'
   */
  namespace?: string,

  /**
   * Path to save generated typeDefs file.
   *
   * Strongly recommend to commit this file to your version control.
   *
   * @default 'src/__generated__/gatsby-types.ts'
   *
   * or (if you use flow)
   * @default 'src/__generated__/gatsby-types.js'
   */
  outputPath?: string,

  /**
   * @deprecated The plugin does not generate useful types for the interface of GatsbyJS schema builder. Will be removed from v4.
   *
   * If you use [schema customization](https://www.gatsbyjs.org/docs/schema-customization/),
   * you might want types for custom resolvers.
   *
   * By default, the plugin only generate types for query data for your components.
   *
   * If set this true, the plugin also generate types for all possible resolver signatures.
   *
   * @default false
   *
   * @see https://graphql-code-generator.com/docs/plugins/typescript-resolvers
   * @see https://graphql-code-generator.com/docs/plugins/flow-resolvers
   */
  includeResolvers?: boolean,

  /**
   * @deprecated Renamed to `autofix`. Will be removed from v4.
   */
  autoFix?: boolean,

  /**
   * Enable auto-fixing `useStaticQuery()` & `<StaticQuery>` with generated type names.
   *
   * @default true
   */
  autofix?: boolean,

  /**
   * Extends or overrides the built-in scalars and custom GraphQL scalars to a custom type.
   *
   * @example
   * ```js
   * {
   *   scalars: {
   *     EmailAddress: 'string',
   *     DateTime: 'number',
   *     URL: 'string',
   *   },
   * }
   * ```
   *
   * @note You cannot override a few GraphQL/GatsbyJS internal scalars.
   * - `Boolean`,
   * - `Buffer`,
   * - `Date`,
   * - `Float`,
   * - `ID`,
   * - `Int`,
   * - `Internal`,
   * - `InternalInput`,
   * - `JSON`,
   * - `Json`,
   * - `Node`,
   * - `NodeInput`,
   * - `Query`,
   * - `String`,
   *
   * @see https://github.com/gatsbyjs/gatsby/blob/6b4b7f81ec/packages/gatsby/src/schema/print.js#L33-L48
   */
  scalars?: { [typename: string]: string },

  /**
   * Emit the Gatsby's schema into the filesystem.
   *
   * Set value as `true` and use default options.
   *
   * @see SchemaOutputOptions
   */
  emitSchema?: {
    [outputPath: string]: boolean | SchemaOutputOptions,
  },

  /**
   * @deprecated Renamed to `emitPluginDocument`. Will be removed from v4.
   *
   * Emit the GraphQL fragments declared by other plugins into the filesystem.
   */
  emitPluginDocuments?: {
    [outputPath: string]: boolean | DocumentOutputOptions,
  },

  /**
   * Emit the GraphQL fragments declared by other plugins into the filesystem.
   */
  emitPluginDocument?: {
    [outputPath: string]: boolean | DocumentOutputOptions,
  },
};

export type DocumentOutputOptions = {
  /**
   * Output file format.
   *
   * By default, the AST will be parsed and printed as GraphQL.
   * If set `json`, the AST will be emitted as JSON.
   *
   * @default 'graphql'
   */
  format: 'graphql' | 'json',
};

export type SchemaOutputOptions = {
  /**
   * Output file format.
   *
   * By default, the introspection result is emitted as JSON.
   * If set `sdl`, introspection will be parsed and printed to GraphQL SDL.
   *
   * @default 'introspection'
   */
  format: 'introspection' | 'sdl',

  /**
   * Include comment descriptions
   *
   * @default true
   */
  commentDescriptions?: boolean,
};

/**
 * Copy of the `IGatsbyResolverContext`
 * @see https://github.com/gatsbyjs/gatsby/blob/f10eb58/packages/gatsby/src/schema/type-definitions.ts
 *
 * @todo Provide more accurate type definition.
 */
export type GatsbyResolverContext = {
  nodeModel: any,
};
