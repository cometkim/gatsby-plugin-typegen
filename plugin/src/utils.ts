import { dirname } from 'path';
import * as fs from 'fs';

export const readFileContent = async (path: string): Promise<string> => {
  return fs.promises.readFile(path, 'utf-8');
};

export const writeFileContent = async (path: string, content: string): Promise<void> => {
  await fs.promises.mkdir(dirname(path), { recursive: true });
  await fs.promises.writeFile(path, content, 'utf-8');
};


