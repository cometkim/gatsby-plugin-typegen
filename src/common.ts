import fs from 'fs';
import { dirname } from 'path';
import { promisify } from 'util';
import { Source } from '@graphql-toolkit/common';

const _mkdir = promisify(fs.mkdir);
const _readFile = promisify(fs.readFile);
const _writeFile = promisify(fs.writeFile);

export const readFile = async (path: string) => _readFile(path, { encoding: 'utf-8' });
export const writeFile = async (path: string, data: string | Buffer) => {
  await _mkdir(dirname(path), { recursive: true });
  await _writeFile(path, data, { encoding: 'utf-8' });
}

export const delay = (ms: number) => new Promise<void>(res => setTimeout(res, ms));

export const formatLanguage = (lang: 'typescript' | 'flow') => (
  (lang === 'typescript') ? 'TypeScript' : 'Flow'
);

export type UnwrapPromise<T> = T extends Promise<infer U> ? U : T;

export function deduplicateFragmentFromDocuments(documents: Source[]) {
  const existFragmentNames = new Set<string>();
  return documents.map(source => {
    const { document } = source;

    // Nothing to do for other sources
    if (!document) {
      return source;
    }

    const uniqDefinitions = document.definitions.filter(def => {
      // De-dup only cares about fragments
      if (def.kind !== 'FragmentDefinition') {
        return true;
      }
      const fragmentName = def.name.value;
      const duplicated = existFragmentNames.has(fragmentName);
      if (!duplicated) existFragmentNames.add(fragmentName);
      return !duplicated;
    });

    return {
      ...source,
      document: {
        ...document,
        definitions: uniqDefinitions,
      },
    };
  });
}
