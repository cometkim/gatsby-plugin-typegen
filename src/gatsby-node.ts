import path from 'path';
import { stripIndent } from 'common-tags';
import { GatsbyNode } from 'gatsby';
import {
  printSchema,
  introspectionFromSchema,
  GraphQLSchema,
} from 'gatsby/graphql';
import FileParser from 'gatsby/dist/query/file-parser';
import { parseGraphQLSDL, Source } from '@graphql-toolkit/common';

import { writeFile, UnwrapPromise, readFile } from './common';
import { setupCodegenWorker, setupInsertTypeWorker, InsertTypeTask } from './workers';
import { requirePluginOptions, RequiredPluginOptions } from './plugin-utils';
import { GatsbyKnownAction } from './gatsby-utils';

// Plugin will track documents what is actually used by Gatsby.
const trackedSource = new Map<string, Source>();

// noop
const noop = () => {};

let pluginOptions: RequiredPluginOptions;
let unsubscribeQueryExtraction: Function;

export const onPreBootstrap: GatsbyNode['onPreBootstrap'] = ({
  store,
  reporter,
}, options) => {
  // Validate plugin options earlier.
  pluginOptions = requirePluginOptions(options, { store, reporter });

  reporter.verbose(
    '[typegen] Successfully validate your configuration.\n'
    + JSON.stringify(pluginOptions, null, 2)
  );
};

export const onPreExtractQueries: GatsbyNode['onPreExtractQueries'] = ({
  store,
  reporter,
}) => {
  reporter.verbose('[typegen] Listen on query extraction');

  const fileParser = new FileParser({ parentSpan: null });

  unsubscribeQueryExtraction = store.subscribe(async () => {
    const lastAction = store.getState().lastAction as GatsbyKnownAction;

    if (lastAction.type !== 'QUERY_EXTRACTION_BABEL_SUCCESS') {
      return;
    }

    const { componentPath } = lastAction.payload;

    if (trackedSource.has(componentPath)) {
      return;
    }

    let extractionResult: UnwrapPromise<ReturnType<typeof fileParser.parseFile>>;
    try {
      extractionResult = await fileParser.parseFile(componentPath, noop);
    } catch {
      extractionResult = null;
    }

    if (extractionResult) {
      const document = parseGraphQLSDL(componentPath, extractionResult.map(doc => doc.text).join(''), {});
      trackedSource.set(componentPath, document);
    }
  });
};

export const onPostBootstrap: GatsbyNode['onPostBootstrap'] = async ({
  store,
  reporter,
}) => {
  const {
    language,
    outputPath,
    includeResolvers,
    emitSchema,
    emitPluginDocuments,
    autoFix,
  } = pluginOptions;

  reporter.verbose('[typegen] End-up listening on query extraction.');
  unsubscribeQueryExtraction();

  const { program, schema: _schema } = store.getState();
  const basePath = program.directory as string;
  const schema = _schema as GraphQLSchema;

  for (const [schemaOutputPath, schemaOutputOptions] of Object.entries(emitSchema)) {
    const { commentDescriptions = true } = schemaOutputOptions;

    let output: string;
    switch (schemaOutputOptions.format) {
      case 'sdl':
        output = printSchema(schema, { commentDescriptions });
        break;
      case 'introspection':
        output = JSON.stringify(introspectionFromSchema(schema, { descriptions: commentDescriptions }), null, 2);
        break;
    }

    reporter.verbose(`[typegen] Emit Gatsby schema into ${schemaOutputPath}`);
    await writeFile(schemaOutputPath, output);
  }

  const pluginDocuments = Object.values(emitPluginDocuments).some(Boolean) && (
    stripIndent(
      Array.from(trackedSource.entries())
        .filter(([componentPath]) => !componentPath.startsWith(path.resolve(basePath, 'src')))
        .map(([, source]) => source.rawSDL)
        .join('\n'),
    )
  );
  if (pluginDocuments) {
    for (const [documentOutputPath, documentOutputOptions] of Object.entries(emitPluginDocuments)) {
      if (!documentOutputOptions) continue;
      reporter.verbose(`[typegen] Emit Gatsby plugin documents into ${documentOutputPath}`);
      await writeFile(path.resolve(basePath, documentOutputPath), pluginDocuments);
    }
  }

  const codegenWorker = setupCodegenWorker({
    schemaAst: schema,
    language,
    outputPath,
    includeResolvers,
    reporter,
  });

  const pushCodegenTask = () => {
    // @ts-ignore
    codegenWorker.push({ documents: [...trackedSource.values()].filter(Boolean) });
  };

  pushCodegenTask();

  if (process.env.NODE_ENV === 'development') {
    reporter.verbose('[typegen][dev] Watching query changes and re-run workers');

    const insertTypeWorker = autoFix && setupInsertTypeWorker({ reporter });
    const pushInsertTypeTask = (task: InsertTypeTask) => {
      if (!insertTypeWorker) {
        return;
      }
      insertTypeWorker.push(task);
    };

    store.subscribe(async () => {
      const lastAction = store.getState().lastAction as GatsbyKnownAction;

      // Listen gatsby actions
      // - QUERY_EXTRACTED action for page queries.
      // - REPLACE_STATIC_QUERY action for static queries.
      if (lastAction.type !== 'QUERY_EXTRACTED' && lastAction.type !== 'REPLACE_STATIC_QUERY') {
        return;
      }

      const { payload: { query, componentPath } } = lastAction;

      const source = trackedSource.get(componentPath);
      if (source?.rawSDL === query) {
        return;
      }

      const document = parseGraphQLSDL(componentPath, query, {
        noLocation: true,
      });
      trackedSource.set(componentPath, document);

      pushCodegenTask();

      if (pluginOptions.language === 'typescript' && /\.tsx?$/.test(componentPath)) {
        pushInsertTypeTask({ file: componentPath });
      }

      // Flow version is bit more slower because should check the `@flow` comment exist.
      if (pluginOptions.language === 'flow' && /\.jsx?$/.test(componentPath)) {
        const content = await readFile(componentPath);
        const hasFlowComment = content.includes('@flow');
        reporter.verbose(`[typegen] Check if the file has flow comment: ${hasFlowComment}`);
        hasFlowComment && pushInsertTypeTask({ file: componentPath });
      }
    });
  }
};
