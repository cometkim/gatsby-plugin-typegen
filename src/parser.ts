import { stripIndent } from 'common-tags';
import { print, parse } from 'gatsby/graphql';
import { readFile } from './common';

const prettifySDL = (value: string) => print(parse(value, { noLocation: true }));

/**
 * @param componentPath path of file to extract static and page queries
 */
export const extractDocuments = (content: string) => {
  return {
    staticQueryComponents: extractStaticQueryComponentDocuments(content),
    staticQueryHooks: extractStaticQueryHookDocuments(content),
    pageQueries: extractPageQueryDocuments(content),
  };
};

/**
 * @param componentPath path of file to get all documents
 */
export const parseFile = async (componentPath: string) => {
  const file = await readFile(componentPath);
  const {
    staticQueryComponents,
    staticQueryHooks,
    pageQueries,
  } = extractDocuments(file);

  if ([
    staticQueryComponents.length === 0,
    staticQueryHooks.length === 0,
    pageQueries.length === 0,
  ].every(Boolean)) {
    return null;
  }

  return staticQueryComponents.join('\n')
    + staticQueryHooks.join('\n')
    + pageQueries.join('\n')
  ;
};

/**
 * Component-style static queries should be:
 * - use Gatsby's StaticQuery component API
 * - use Gatsby's graphql tagged template literal
 *
 * (?<JsxTagOpening><StaticQuery
 *   (?<TagTypeTemplate><
 *     (?<TagTypeArgument>\S+)
 *   >)?
 * )
 * [\s\S]+?
 * query={
 * [\s\S]*?
 * graphql
 * (?<TemplateLiteral>`\s*
 *   (?<QueryDefinitionStart>
 *     query\s+(?<QueryName>\S*)
 *     \s*\{
 *   )
 *   [^`]*?
 * `)
 */
const STATIC_QUERY_COMPONENT_REGEXP = /(?<JsxTagOpening><StaticQuery(?<TagTypeTemplate><(?<TagTypeArgument>\S+)>)?)[\s\S]+?query={[\s\S]*?graphql(?<TemplateLiteral>`\s*(?<QueryDefinition>(?<QueryDefinitionStart>query\s+(?<QueryName>\S*)\s*\{)[^`]*)`)/g;


export const extractStaticQueryComponentDocuments = (content: string): string[] => {
  const result: string[] = [];

  let match: RegExpExecArray | null = null;
  while(match = STATIC_QUERY_COMPONENT_REGEXP.exec(content)) {
    if (!(match && match.groups)) {
      continue;
    }
    let rawSDL = match.groups['QueryDefinition'];
    try {
      rawSDL = prettifySDL(rawSDL);
    } catch { /* drop error */ }
    result.push(rawSDL);
    break;
  }

  return result;
};

/**
 * Hook-style static queries should be:
 * - use Gatsby's useStaticQuery hook API
 * - use Gatsby's graphql tagged template literal
 *
 * (?<CallExpressionName>useStaticQuery
 *   (?<TypeTemplate><
 *     (?<TypeArgument>\S*)
 *   >)?
 * )
 * \([\s\S]*?
 * graphql
 * (?<TemplateLiteral>`\s*
 *   (?<QueryDefinition>
 *     (?<QueryDefinitionStart>
 *       query\s+(?<QueryName>\S*)
 *       \s*\{
 *     )
 *     [^`]*
 *   )
 * `)
 */
const STATIC_QUERY_HOOK_REGEXP = /(?<CallExpressionName>useStaticQuery(?<TypeTemplate><(?<TypeArgument>\S*)>)?)\([\s\S]*?graphql(?<TemplateLiteral>`\s*(?<QueryDefinition>(?<QueryDefinitionStart>query\s+(?<QueryName>\S*)\s*\{)[^`]*)`)/g;

export const extractStaticQueryHookDocuments = (content: string) => {
  const result: string[] = [];

  let match: RegExpExecArray | null = null;
  while(match = STATIC_QUERY_HOOK_REGEXP.exec(content)) {
    if (!(match && match.groups)) {
      continue;
    }
    let rawSDL = match.groups['QueryDefinition'];
    try {
      rawSDL = prettifySDL(rawSDL);
      result.push(rawSDL);
    } catch { /* drop error */ }
  }

  return result;
};

/**
 * Page queries should be:
 * - valid var|let|const binding
 * - exported
 * - use Gatsby's graphql tagged template literal
 * - only a single page query allowed (use first one)
 *
 * export
 * \s+
 * (var|let|const)
 * \s+
 * (?<VariableName>\S+)
 * \s*
 * =
 * \s*
 * graphql
 * (?<TemplateLiteral>`\s*
 *   (?<QueryDefinitionStart>
 *     query\s+(?<QueryName>\S*)
 *     \s*\{
 *   )
 *   [^`]*?
 * `)
 */
const PAGE_QUERY_REGEXP = /export\s+(var|let|const)\s+(?<VariableName>\S+)\s*=\s*graphql(?<TemplateLiteral>`\s*(?<QueryDefinition>(?<QueryDefinitionStart>query\s+(?<QueryName>\S*)\s*\{)[^`]*)`)/;

export const extractPageQueryDocuments = (content: string) => {
  const result: string[] = [];

  let match = PAGE_QUERY_REGEXP.exec(content);
  if (!(match && match.groups)) {
    return result;
  }
  let rawSDL = match.groups['QueryDefinition'];
  try {
    rawSDL = prettifySDL(rawSDL);
    result.push(rawSDL);
  } catch { /* drop error */ }

  return result;
};
