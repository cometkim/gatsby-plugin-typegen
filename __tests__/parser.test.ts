import path from 'path';
import { readFile } from '../src/common';
import { extractStaticQueryComponentDocuments, extractStaticQueryHookDocuments, extractPageQueryDocuments, parseFile } from '../src/parser';

describe('parser', () => {
  describe('extractStaticQueryComponentDocuments', () => {
    it('extract the document properly', async () => {
      const file = await readFile(path.resolve(__dirname, './data/static-query-component.tsx'));
      const documents = await extractStaticQueryComponentDocuments(file);
      expect(documents).toEqual([
`query Heading {
  site {
    siteMetadata {
      title
    }
  }
}
`
      ]);
    });
  });

  describe('extractStaticQueryHookDocuments', () => {
    it('extract documents properly', async () => {
      const file = await readFile(path.resolve(__dirname, './data/static-query-hook.tsx'));
      const documents = await extractStaticQueryHookDocuments(file);
      expect(documents).toEqual([
`query Image1 {
  placeholderImage: file(relativePath: {eq: "gatsby-astronaut.png"}) {
    childImageSharp {
      fluid(maxWidth: 300) {
        ...GatsbyImageSharpFluid
      }
    }
  }
}
`,
`query Image2 {
  placeholderImage: file(relativePath: {eq: "gatsby-astronaut.png"}) {
    childImageSharp {
      fluid(maxWidth: 300) {
        ...GatsbyImageSharpFluid
      }
    }
  }
}
`,
`query Image3 {
  placeholderImage: file(relativePath: {eq: "gatsby-astronaut.png"}) {
    childImageSharp {
      fluid(maxWidth: 300) {
        ...GatsbyImageSharpFluid
      }
    }
  }
}
`,
      ]);
    });
  });

  describe('extractPageQueryDocuments', () => {
    it('extract documents properly', async () => {
      const file = await readFile(path.resolve(__dirname, './data/page-query.tsx'));
      const documents = extractPageQueryDocuments(file);
      expect(documents).toHaveLength(1);
      expect(documents).toEqual([
`query HomePage1 {
  site {
    siteMetadata {
      description
    }
  }
}
`,
      ]);
    });
  });

  describe('parseFile', () => {
    it('extract all documents string from the file', async () => {
      const documents = await parseFile(path.resolve(__dirname, './data/many-documents.jsx'));
      expect(documents).toMatchSnapshot();
    });
  });
});
