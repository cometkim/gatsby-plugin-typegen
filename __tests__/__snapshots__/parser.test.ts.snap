// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser parseFile extract all documents string from the file 1`] = `
"query SiteMetadata {
  site {
    siteMetadata {
      title
      description
    }
  }
}
query IndexPageTemplate {
  markdownRemark(frontmatter: {templateKey: {eq: \\"index-page\\"}}) {
    frontmatter {
      title
      subtitle
      image {
        childImageSharp {
          fluid(maxWidth: 2048, quality: 100) {
            ...GatsbyImageSharpFluid
          }
        }
      }
      heading
      mainpitch {
        title
        description
      }
      bigimage {
        image {
          childImageSharp {
            fluid(maxWidth: 240, quality: 64) {
              ...GatsbyImageSharpFluid
            }
          }
          publicURL
        }
        alt
      }
      intro {
        blurbs {
          image {
            childImageSharp {
              fluid(maxWidth: 240, quality: 64) {
                ...GatsbyImageSharpFluid
              }
            }
          }
          text
        }
        heading
        description
      }
    }
  }
  allMarkdownRemark(sort: {order: DESC, fields: [frontmatter___featuredpost, frontmatter___date]}, filter: {frontmatter: {templateKey: {eq: \\"blog-post\\"}}}, limit: 1) {
    edges {
      node {
        excerpt(pruneLength: 400)
        id
        fields {
          slug
        }
        frontmatter {
          title
          templateKey
          date(formatString: \\"MMMM DD, YYYY\\")
          featuredpost
          featuredimage {
            childImageSharp {
              fluid(maxWidth: 120, quality: 100) {
                ...GatsbyImageSharpFluid
              }
            }
          }
        }
      }
    }
  }
}
"
`;
